### PR TITLE
Aqaint 181

### PR DESCRIPTION
**Pull request by vkrasnobelmov**
Registration. Feature file.

### CODE REVIEW

File: BookingSpecTests/RegistrationFeature.feature [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+Scenario: Input valid credentials in order to Register
+	When I open browser
+	And I navigate to url "https://booking.com"
+	And I click Register button on Ribbon menu
+	And I set following parameters on Register Pop Up dialog
+	| Field    | Value                      |
+	| Email    | testmirantistest@gmail.com |
+	| Password | 1234qweR                   |
+	And I click Get Started button on PopUp
+	Then I see that I am Registered and Signed In
```
Comment (IgorRukhlenko) [inline BookingSpecTests/RegistrationFeature.feature]: Why we cannot use step 'Then I see that I am Signed In ' here?

File: BookingSpecTests/RegistrationFeature.feature [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	Scenarios: 
+		| Email        | Password  | Message                                         |
+		| 2131312@.com | 1         | Password needs to be at least 8 characters long |
+		| 2131312@.com |           | Please add a password                           |
+		| 2131312@.com | wrongpass | Please enter a valid email address.             |
+		| 2131312.com  | 12341234  | Please enter a valid email address.             |
+		| 2131312@com  | 12341234  | Please enter a valid email address.             |
+		|              |           | Please enter a valid email address.             |
+
+Scenario: Input email and pass that already registered and make sure that I am signed in
```
Comment (IgorRukhlenko) [inline BookingSpecTests/RegistrationFeature.feature]: Let's rename it to something more simple like: Try to register as already registered user.

File: BookingSpecTests/RegistrationFeature.feature [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	And I navigate to url "https://booking.com"
+	And I click Register button on Ribbon menu
+	And I set following parameters on Register Pop Up dialog
+	| Field    | Value                      |
+	| Email    | testmirantistest@gmail.com |
+	| Password | 1234qweR                   |
+	And I click Get Started button on PopUp
+	Then I see that I am Signed In
+
+Scenario: Input email that already registered and invalid pass to make sure that I am not registered and not signed in
```
Comment (IgorRukhlenko) [inline BookingSpecTests/RegistrationFeature.feature]: Let's make the scenario name more simple and clear.

File: BookingSpecTests/RegistrationFeature.feature [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	When I open browser
+	And I navigate to url "https://booking.com"
+	And I click Register button on Ribbon menu
+	And I set following parameters on Register Pop Up dialog
+	| Field    | Value                      |
+	| Email    | testmirantistest@gmail.com |
+	| Password | 12341234                   |
+	And I click Get Started button on PopUp
+	Then I see that I am not Signed In and not Registered
```
Comment (IgorRukhlenko) [inline BookingSpecTests/RegistrationFeature.feature]: Not obvious step. Let's split it.

File: BookingSpecTests/RegistrationFeature.feature [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	When I open browser
+	And I navigate to url "https://booking.com"
+	And I click Register button on Ribbon menu
+	And I set following parameters on Register Pop Up dialog
+		| Field    | Value                      |
+		| Email    | testmirantistest@gmail.com |
+		| Password | 1234qweR                   |
+	And I click on X button
+	And I click Register button on Ribbon menu
+	Then I see that credential values are saved in Register PopUp fields
```
Comment (IgorRukhlenko) [inline BookingSpecTests/RegistrationFeature.feature]: How are you going to check that proper credentials are saved?

File: BookingSpecTests/RegistrationFeature.feature [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	And I navigate to url "https://booking.com"
+	And I click Register button on Ribbon menu
+	And I set following parameters on Register Pop Up dialog
+		| Field    | Value                      |
+		| Email    | testmirantistest@gmail.com |
+		| Password | 1234qwER                   |
+	And I click on X button
+	And I refresh page
+	And  I click Register button on Ribbon menu
+	Then I see that credential values are not saved in Register PopUp fields
```
Comment (IgorRukhlenko) [inline BookingSpecTests/RegistrationFeature.feature]: How are you going to check that proper credentials are not saved?

File: BookingSpecBindings/Bindings/SignInPopUpBindings.cs [hunk 7]
```diff
@@ -1,29 +1,85 @@
...
+		public void WhenIClickForgotYourPasswordButton()
+		{
+			signPage.ClickForgotPass();
+		}
+		[Then(@"I see error message ""(.*)"" on Sign In PopUp")]
+		public void ThenISeeErrorMessageOnSignInPopUp(string error)
+		{
+			Assert.AreEqual(signPage.GetErrorText(), error);
+		}
+		[Then(@"I see that credential values are saved in Sign In PopUp fields")]
```
Comment (IgorRukhlenko) [inline BookingSpecBindings/Bindings/SignInPopUpBindings.cs]: Such step implementation cannot be used for other users and there is no check that password also saved.

File: BookingSpecBindings/Bindings/SignInPopUpBindings.cs [hunk 8]
```diff
@@ -34,3 +90,16 @@ public void WhenIWriteEmailOnSignInPopUp(string email)
 		{
 			signPage.TypeEmail(email);
 		}
+		[Then(@"I see Sign In PopUp dialog")]
+		public bool ThenISeeSignInPopUpDialog()
+		{
+			Console.WriteLine(signPage.PopUpButton.Displayed);
+			return signPage.PopUpButton.Displayed;
+		}
+		[Then(@"I see that credential values are not saved in Sign In PopUp fields")]
```
Comment (IgorRukhlenko) [inline BookingSpecBindings/Bindings/SignInPopUpBindings.cs]: There is no check for password field.

**Review by IgorRukhlenko — CHANGES_REQUESTED** [review]

**Review by IgorRukhlenko — APPROVED** [review]

File: BookingSpecBindings/Bindings/SignInPopUpBindings.cs [hunk 9]
```diff
@@ -96,10 +104,25 @@ public bool ThenISeeSignInPopUpDialog()
 			Console.WriteLine(signPage.PopUpButton.Displayed);
 			return signPage.PopUpButton.Displayed;
 		}
-		[Then(@"I see that credential values are not saved in Sign In PopUp fields")]
-		public void ThenISeeThatCredentialValuesAreNotSavedInSignInPopUpFields()
+		[Then(@"I check that I am not Registered")]
```
Comment (etolmachev) [inline BookingSpecBindings/Bindings/SignInPopUpBindings.cs]: Step name doesn't apply to what it does. Check it please

File: BookingSpecBindings/TestBase/Pages/RegisterPopUp.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+		{
+			XmarkButton.Click();
+		}
+		public string GetErrorText()
+		{
+			return new HtmlElement(By.CssSelector(MsgErrorLocator)).Text;
+		}
+		public string GetEmailText()
+		{
+			string EmailValue = EmailField.GetAttribute("value");
```
Comment (etolmachev) [inline BookingSpecBindings/TestBase/Pages/RegisterPopUp.cs]: Make it one line method to be consistent

File: BookingSpecBindings/TestBase/Pages/RegisterPopUp.cs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+			return EmailValue;
+		}
+		public string GetPassText()
+		{
+			string PassValue = PassField.GetAttribute("value");
```
Comment (etolmachev) [inline BookingSpecBindings/TestBase/Pages/RegisterPopUp.cs]: Make it one line method to be consistent

File: BookingSpecBindings/TestBase/Pages/SignInPopUp.cs [hunk 12]
```diff
@@ -66,11 +65,15 @@ public string GetEmailText()
 			string EmailValue = EmailField.GetAttribute("value");
```
Comment (etolmachev) [inline BookingSpecBindings/TestBase/Pages/SignInPopUp.cs]: Make it one line method to be consistent

File: BookingSpecBindings/TestBase/Pages/SignInPopUp.cs [hunk 13]
```diff
@@ -66,11 +65,15 @@ public string GetEmailText()
...
 			return EmailValue;
 		}
+		public string GetPassText()
+		{
+			string PassValue = PassField.GetAttribute("value");
```
Comment (etolmachev) [inline BookingSpecBindings/TestBase/Pages/SignInPopUp.cs]: Make it one line method to be consistent

File: BookingSpecBindings/Utils.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using TechTalk.SpecFlow;
+
+namespace BookingSpecBindings
+{
+
+	class Utils
```
Comment (etolmachev) [inline BookingSpecBindings/Utils.cs]: This class must be static.

File: BookingSpecBindings/Utils.cs [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+	{
+		private static Random random = new Random();
+		public static string rnd(string email)
```
Comment (etolmachev) [inline BookingSpecBindings/Utils.cs]: It's better to pass to rnd() only int length. This method should not be responsible for parsing strings.

File: BookingSpecBindings/Utils.cs [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+			string mid = new string(Enumerable.Repeat(chars, length)
+			.Select(s => s[random.Next(s.Length)]).ToArray());
+			string left = email.Substring(0, email.IndexOf("{{"));
+			string right = email.Substring(email.IndexOf("}}") + 2);
+			return left + mid + right;
+		}
+		private static string regexAll = "(?={{)(.*?)(?<=}})";
+		private static string regexLeft = "(?<={{)(.*?)(?=::)";
+		private static string regexRight = "(?<=::)(.*?)(?=}})";
+		public static string Context(string email)
```
Comment (etolmachev) [inline BookingSpecBindings/Utils.cs]: You should pass key right away, this method is about working with context, not about parsing strings.

File: BookingSpecBindings/Utils.cs [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using TechTalk.SpecFlow;
+
+namespace BookingSpecBindings
+{
+
+	class Utils
+	{
+		private static Random random = new Random();
```
Comment (etolmachev) [inline BookingSpecBindings/Utils.cs]: Must be defined in rnd method

File: BookingSpecBindings/Utils.cs [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+		}
+		private static string regexAll = "(?={{)(.*?)(?<=}})";
+		private static string regexLeft = "(?<={{)(.*?)(?=::)";
+		private static string regexRight = "(?<=::)(.*?)(?=}})";
+		public static string Context(string email)
+		{
+			string key = Regex.Match(email, regexRight).ToString();
+			return ScenarioContext.Current.Get<string>(key);
+		}
+		public static string Resolve(string result)
```
Comment (etolmachev) [inline BookingSpecBindings/Utils.cs]: Rename result to something more suitable.

File: BookingSpecBindings/Utils.cs [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+		{
+			string placeholder = Regex.Match(result, regexLeft).ToString();
```
Comment (etolmachev) [inline BookingSpecBindings/Utils.cs]: Rename placeholder to "methodName"

File: BookingSpecBindings/Utils.cs [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+			Console.WriteLine(placeholder);
```
Comment (etolmachev) [inline BookingSpecBindings/Utils.cs]: Remove unneeded debug output

File: BookingSpecBindings/Utils.cs [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+			switch (placeholder)
+			{
+				case "rnd":
+					return rnd(result);
```
Comment (etolmachev) [inline BookingSpecBindings/Utils.cs]: It's better for rnd and context methods return only resolved value for {{a::b}}. You should combine final result string in Resolve method.

**Review by etolmachev — CHANGES_REQUESTED** [review]

**Pull request merged**